pkg/fetcher/fetcher.go

```diff
@@ -414,6 +449,44 @@ func (u *URLFetcher) IsNonretryableClientError() bool {
 		s != http.StatusLocked && s != http.StatusTooManyRequests
 }
 
+// Extract message from error response body (#5951)
+// The JSON format follows the convention of Docker
+func (u *URLFetcher) extractErrResponseMessage(rdr io.ReadCloser) (string, error) {
+	out := bytes.NewBuffer(nil)
+	_, err := io.Copy(out, rdr)
+	if err != nil {
+		log.Debugf("Error when copying from error response body stream: %s", err)
+		return "", err
+	}
+
+	res := []byte(out.Bytes())
+	log.Debugf("Error message json string: %s", string(res))
+
+	var errResponse ErrorResponseBody
+	err = json.Unmarshal(res, &errResponse)
+	if err != nil {
+		log.Debugf("Error when unmarshalling error response body: %s", err)
+		return "", err
+	}
+
+	if errResponse.Errors == nil || len(errResponse.Errors) == 0 {
+		log.Debugf("Error response wrong format. Response body: %s", string(res))
+		return "", fmt.Errorf("Error response wrong format")
+	}
+
+	// grab out every error message
+	var errString string
```

Comment:
This can be made simpler and cleaner with

```
for i, error := range errResponse.Errors {
  if i > 0 {
    errString += ", "
  }
  errString += error.Message
}
```

They're equivalent, but I find for range cleaner than traditional C style for loops.